src/datachain/client/gcs.py

```diff
@@ -32,6 +32,14 @@ def create_fs(cls, **kwargs) -> GCSFileSystem:
 
         return cast(GCSFileSystem, super().create_fs(**kwargs))
 
+    def url(self, path: str, expires: int = 3600, **kwargs) -> str:
+        try:
+            return self.fs.sign(self.get_full_path(path), expiration=expires, **kwargs)
+        except AttributeError as exc:
+            if "you need a private key to sign credentials" in str(exc):
```

Comment:
Is there really this message in `AttributeError`, and what raises this?

Could you please share stacktrace?

Reply:
Sure!

```
In [1]: from datachain.catalog import get_catalog

In [2]: catalog = get_catalog()

In [3]: url = catalog.signed_url('gs://datachain-demo', 'aspset510/test/videos/1e28/1e28-0026-right.mp4', {'anon': True})
> /Users/vlad/work/iterative/datachain/src/datachain/client/gcs.py(40)url()
     39             import ipdb; ipdb.set_trace()
---> 40             print(exc)
     41

ipdb> exc
AttributeError("you need a private key to sign credentials.the credentials you are currently using <class 'google.oauth2.credentials.Credentials'> just contains a token. see https://googleapis.dev/python/google-api-core/latest/auth.html#setting-up-a-service-account for more details.")
```

Reply:
Where is this exception is being raised? The above stacktrace is just your `print(exc)` function.

Reply:
Sorry, it was wrong code I was copied. Updated the stacktrace.

Reply:
Also, what is the expectation here for `url()` method? Should it return `https://storage.googleapis.com/...` url when it cannot create a signed url, or should that only be the case for public bucket?

If the latter, how can we be sure that this exception only gets raised for public buckets? (sorry too many questions).

Reply:
These are good question, please, feel free to ask any 🤗
What if we will return public URL only if [`anon` param](https://github.com/iterative/datachain/blob/main/src/datachain/client/gcs.py#L30-L31) is set? 🤔

Reply:
Like this, for example, should be more accurate:
```diff
diff --git a/src/datachain/client/gcs.py b/src/datachain/client/gcs.py
index 99e3c14..6eaaf1c 100644
--- a/src/datachain/client/gcs.py
+++ b/src/datachain/client/gcs.py
@@ -36,7 +36,8 @@ class GCSClient(Client):
         try:
             return self.fs.sign(self.get_full_path(path), expiration=expires, **kwargs)
         except AttributeError as exc:
-            if "you need a private key to sign credentials" in str(exc):
+            is_anon = self.fs.storage_options.get("token") == "anon"
+            if is_anon and "you need a private key to sign credentials" in str(exc):
                 return f"https://storage.googleapis.com/{self.name}/{path}"
             raise
```

Reply:
Or may be even like this?
```python
    def url(self, path: str, expires: int = 3600, **kwargs) -> str:
        if self.fs.storage_options.get("token") == "anon":
            return f"https://storage.googleapis.com/{self.name}/{path}"
        return self.fs.sign(self.get_full_path(path), expiration=expires, **kwargs)
```

Reply:
Choose the first option just to be on the safe side and to do not break any existing flows. Although second option looks like a better one.